src/browser/history.js

```diff
@@ -0,0 +1,320 @@
+      const data = yield db.get(record[PouchDB.id]);
+      return new record.constructor(data);
+    } catch (error) {
+      if (error.status != 404) {
+        throw error
+      }
+      return record;
+    }
+  });
+
+  const edit = async(function*(record, transform, db, config) {
+    const current = yield read(record, db);
+    const next = yield transform(current);
+    return write(next, db, config);
+  });
+
+  const push = x => xs => xs.push(x);
+  const remove = x => xs => xs.remove(x);
+
+  const exclude = x => xs => {
+    const index = xs.indexOf(x)
+    return index < 0 ? xs :
+           index == 0 ? xs.rest() :
+           index + 1 == xs.size ? xs.butLast() :
+           xs.take(index).concat(xs.skip(index + 1))
+  }
+  const include = x => xs => xs.indexOf(x) < 0 ? xs.push(x) : xs
+
+
+  /*
+  ## Interests API
+
+  ### Sites
+
+  Sites store contains records of sites that have being visited by a user.
+  Sites store contains records of the following schema (which could be extended
+  in the future).
+
+  {
+    _id: "site/http://learnyouahaskell.com/introduction#about-this-tutorial",
+    uri: "http://learnyouahaskell.com/introduction#about-this-tutorial",
+    title: "Introduction - Learn You a Haskell for Great Good!",
+    backgroundColor: "rgb(255, 255, 255)",
+    visits: [
+      {
+        start: 1421434329682266,
+        end: 1421434484899,
+        device: "Desktop"
+      }
+    ],
+    tags: [
+      "haskell",
+      "functional"
+    ]
+  }
+  */
+
+  const Visit = Record({
+    type: Type('Visit'),
+    start: TimeStamp,
+    end: Maybe(TimeStamp),
+    id: String, // This is a web view ID in our case.
+    device: String('Desktop'),
+  });
+
+  const Site = Record({
+    [PouchDB.id]: ID,
+    [PouchDB.revision]: Maybe(Revision),
+    type: Type('Site'),
+    uri: URI,
+    title: Maybe(String),
+    visits: List(Visit),
+    tags: List(TagName),
+    icon: Maybe(URI), // Would be better if it was Blob as well.
+    image: Blob,
+  });
+  Site.storeID = 'sites';
+  Site.frequency = ({visits}) => visits.size;
+  Site.from = ({uri, title}) => Site({[PouchDB.id]: `site/${uri}`, uri, title});
+
+  Site.beginVisit = ({time, id, device}) => site =>
+    site.update('visits', push(Visit({start: time, id, device})));
+
+
+  Site.endVisit = ({id, time}) => site => {
+    const index = site.visits.findIndex(visit => visit.id === id);
+    return index < 0 ? site : site.setIn(['visits', index, 'end'], time);
+  };
+
+
+/**
+
+### Quotes
+
+Quotes store contains records of quotes that have being created by a user that
+have a following structure.
+
+
+{
+  _id: "quote/W29iamVjdCBBcnJheUJ1ZmZlcl0=",
+  content: `If you say that <span class="fixed">a</span> is 5, you can't say it's something else later because you just said it was 5. What are you, some kind of liar? So in purely functional languages, a function has no side-effects. The only thing a function can do is calculate something and return it as a result.`
```

Comment:
Can we leave it as is for now as it would really depend on implementatiot details, like image can be captures as base64 uris for example or we may need to capture styles as well, etc...

I'm happy to change this and rename once we start worknig on it. I just included it so I don't loose implementatino that I did back when it was `firefox.html`.